File: 2017/go/day1/day01.go
package main

import (
	"fmt"
	"io/ioutil"
	"strconv"
	"strings"
)

func Sum(input string) int {
	sum := 0
	// Account for circular list
	if input[0] == input[len(input)-1] {
		digit, err := strconv.Atoi(string(input[0]))
		if err == nil {
			sum += digit
		}
	}
	current, input := input[0], input[1:]
	for len(input) > 0 {
		if current == input[0] {
			digit, err := strconv.Atoi(string(current))
			if err == nil {
				sum += digit
			}
		}
		current, input = input[0], input[1:]
	}

	return sum
}

func HalfSum(input string) int {
	sum := 0
	offset := len(input) / 2
	half := input[:offset]
	for i, d := range half {
		if byte(d) == input[i+offset] {
			digit, err := strconv.Atoi(string(d))
			if err == nil {
				sum += 2 * digit
			}
		}
	}

	return sum
}

func main() {
	digits, err := ioutil.ReadFile("./day01.in")
	if err == nil {
		input := strings.TrimSpace(string(digits))
		fmt.Printf("Day 1 Part 1: %+v\n", Sum(input))
		fmt.Printf("Day 1 Part 2: %+v\n", HalfSum(input))
	}
}

File: 2017/day01.rb
def sum input
  digits = input.strip.chars.map { |x| x.to_i }
  sum = 0

  # Account for list of digits being circular
  sum += digits.first if (digits.last == digits.first)

  current = digits.shift

  while !current.nil?
    sum += current if current == digits.first
    current = digits.shift
  end

  sum
end

def half_sum input
  digits = input.strip.chars.map { |x| x.to_i }
  offset = digits.size/2
  sum = 0
  digits.each_with_index do |d, i|
    if d == digits[i + offset]
      sum += 2*d
    end
  end
  sum
end

if $0 == __FILE__
  input = File.read('day01.in')
  p sum input
  p half_sum input
end

